// @mat-datetimepicker/core/esm2015/datetimepicker/datetimepicker.module.js
import { A11yModule } from "@angular/cdk/a11y";
import { OverlayModule } from "@angular/cdk/overlay";
import { CommonModule } from "@angular/common";
import { NgModule } from "@angular/core";
import { MatButtonModule } from "@angular/material/button";
import { MatDialogModule } from "@angular/material/dialog";
import { MatIconModule } from "@angular/material/icon";
import { MatDatetimepickerCalendar } from "./calendar";
import { MatDatetimepickerCalendarBody } from "./calendar-body";
import { MatDatetimepickerClock } from "./clock";
import { MatDatetimepicker, MatDatetimepickerContent } from "./datetimepicker";
import { MatDatetimepickerInput } from "./datetimepicker-input";
import { MatDatetimepickerToggle } from "./datetimepicker-toggle";
import { MatDatetimepickerMonthView } from "./month-view";
import { MatDatetimepickerYearView } from "./year-view";
import { MatDatetimepickerMultiYearView } from "./multi-year-view";
import * as ɵngcc0 from '@angular/core';
export class MatDatetimepickerModule {
}
MatDatetimepickerModule.ɵfac = function MatDatetimepickerModule_Factory(t) { return new (t || MatDatetimepickerModule)(); };
MatDatetimepickerModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: MatDatetimepickerModule });
MatDatetimepickerModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            MatButtonModule,
            MatDialogModule,
            MatIconModule,
            OverlayModule,
            A11yModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MatDatetimepickerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    MatButtonModule,
                    MatDialogModule,
                    MatIconModule,
                    OverlayModule,
                    A11yModule
                ],
                entryComponents: [
                    MatDatetimepickerContent
                ],
                declarations: [
                    MatDatetimepickerCalendar,
                    MatDatetimepickerCalendarBody,
                    MatDatetimepickerClock,
                    MatDatetimepicker,
                    MatDatetimepickerToggle,
                    MatDatetimepickerInput,
                    MatDatetimepickerContent,
                    MatDatetimepickerMonthView,
                    MatDatetimepickerYearView,
                    MatDatetimepickerMultiYearView
                ],
                exports: [
                    MatDatetimepickerCalendar,
                    MatDatetimepickerCalendarBody,
                    MatDatetimepickerClock,
                    MatDatetimepicker,
                    MatDatetimepickerToggle,
                    MatDatetimepickerInput,
                    MatDatetimepickerContent,
                    MatDatetimepickerMonthView,
                    MatDatetimepickerYearView,
                    MatDatetimepickerMultiYearView
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MatDatetimepickerModule, { declarations: function () { return [MatDatetimepickerCalendar,
        MatDatetimepickerCalendarBody,
        MatDatetimepickerClock,
        MatDatetimepicker,
        MatDatetimepickerToggle,
        MatDatetimepickerInput,
        MatDatetimepickerContent,
        MatDatetimepickerMonthView,
        MatDatetimepickerYearView,
        MatDatetimepickerMultiYearView]; }, imports: function () { return [CommonModule,
        MatButtonModule,
        MatDialogModule,
        MatIconModule,
        OverlayModule,
        A11yModule]; }, exports: function () { return [MatDatetimepickerCalendar,
        MatDatetimepickerCalendarBody,
        MatDatetimepickerClock,
        MatDatetimepicker,
        MatDatetimepickerToggle,
        MatDatetimepickerInput,
        MatDatetimepickerContent,
        MatDatetimepickerMonthView,
        MatDatetimepickerYearView,
        MatDatetimepickerMultiYearView]; } }); })();
