// @ionic-native/file-opener/ngx/bundle.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib = require('tslib');
var core$1 = require('@angular/core');
var core = require('@ionic-native/core');

var ɵngcc0 = require('@angular/core');
var FileOpener = /** @class */ (function (_super) {
    tslib.__extends(FileOpener, _super);
    function FileOpener() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    FileOpener.prototype.open = function (filePath, fileMIMEType) { return core.cordova(this, "open", { "callbackStyle": "object", "successName": "success", "errorName": "error" }, arguments); };
    FileOpener.prototype.uninstall = function (packageId) { return core.cordova(this, "uninstall", { "callbackStyle": "object", "successName": "success", "errorName": "error" }, arguments); };
    FileOpener.prototype.appIsInstalled = function (packageId) { return core.cordova(this, "appIsInstalled", { "callbackStyle": "object", "successName": "success", "errorName": "error" }, arguments); };
    FileOpener.prototype.showOpenWithDialog = function (filePath, fileMIMEType) { return core.cordova(this, "showOpenWithDialog", { "callbackStyle": "object", "successName": "success", "errorName": "error" }, arguments); };
    FileOpener.pluginName = "FileOpener";
    FileOpener.plugin = "cordova-plugin-file-opener2";
    FileOpener.pluginRef = "cordova.plugins.fileOpener2";
    FileOpener.repo = "https://github.com/pwlin/cordova-plugin-file-opener2";
    FileOpener.platforms = ["Android", "iOS", "Windows", "Windows Phone 8"];
FileOpener.ɵfac = /*@__PURE__*/ function () { var ɵFileOpener_BaseFactory; return function FileOpener_Factory(t) { return (ɵFileOpener_BaseFactory || (ɵFileOpener_BaseFactory = ɵngcc0.ɵɵgetInheritedFactory(FileOpener)))(t || FileOpener); }; }();
FileOpener.ɵprov = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjectable({ token: FileOpener, factory: function (t) { return FileOpener.ɵfac(t); } });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(FileOpener, [{
        type: core$1.Injectable
    }], null, null); })();
    return FileOpener;
}(core.IonicNativePlugin));

exports.FileOpener = FileOpener;
