// @mat-datetimepicker/core/esm2015/adapter/adapter.module.js
import { NgModule } from "@angular/core";
import { MatNativeDateModule, NativeDateModule } from "@angular/material/core";
import { DatetimeAdapter } from "./datetime-adapter";
import { MAT_DATETIME_FORMATS } from "./datetime-formats";
import { NativeDatetimeAdapter } from "./native-datetime-adapter";
import { MAT_NATIVE_DATETIME_FORMATS } from "./native-datetime-formats";
// tslint:disable max-classes-per-file
import * as ɵngcc0 from '@angular/core';
export class NativeDatetimeModule {
}
NativeDatetimeModule.ɵfac = function NativeDatetimeModule_Factory(t) { return new (t || NativeDatetimeModule)(); };
NativeDatetimeModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NativeDatetimeModule });
NativeDatetimeModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [
        {
            provide: DatetimeAdapter,
            useClass: NativeDatetimeAdapter
        }
    ], imports: [[NativeDateModule]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NativeDatetimeModule, [{
        type: NgModule,
        args: [{
                imports: [NativeDateModule],
                providers: [
                    {
                        provide: DatetimeAdapter,
                        useClass: NativeDatetimeAdapter
                    }
                ]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NativeDatetimeModule, { imports: function () { return [NativeDateModule]; } }); })();
const ɵ0 = MAT_NATIVE_DATETIME_FORMATS;
export class MatNativeDatetimeModule {
}
MatNativeDatetimeModule.ɵfac = function MatNativeDatetimeModule_Factory(t) { return new (t || MatNativeDatetimeModule)(); };
MatNativeDatetimeModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: MatNativeDatetimeModule });
MatNativeDatetimeModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ providers: [{ provide: MAT_DATETIME_FORMATS, useValue: ɵ0 }], imports: [[
            NativeDatetimeModule,
            MatNativeDateModule
        ]] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(MatNativeDatetimeModule, [{
        type: NgModule,
        args: [{
                imports: [
                    NativeDatetimeModule,
                    MatNativeDateModule
                ],
                providers: [{ provide: MAT_DATETIME_FORMATS, useValue: ɵ0 }]
            }]
    }], null, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(MatNativeDatetimeModule, { imports: function () { return [NativeDatetimeModule, MatNativeDateModule]; } }); })();
export { ɵ0 };
